File: crates/gravity/src/octree.rs
use std::sync::OnceLock;
use std::time::{Duration, Instant};

use crate::kernel::KernelKind;
use crate::multipole::{translate_multipole, MultipoleMoment, MultipoleMoments};

// ---------------------------------------------------------------------------
// Timing helpers (only active when GRAVITY_TIMING env var is set)
// ---------------------------------------------------------------------------

#[inline]
pub(crate) fn timing_enabled() -> bool {
    static ENABLED: OnceLock<bool> = OnceLock::new();
    *ENABLED.get_or_init(|| {
        std::env::var("GRAVITY_TIMING")
            .map(|v| {
                let v = v.trim();
                !(v.is_empty() || v == "0" || v.eq_ignore_ascii_case("false"))
            })
            .unwrap_or(false)
    })
}

#[inline]
pub(crate) fn log_timing(label: &str, dt: Duration) {
    eprintln!("[gravity-timing] {label}: {:.3} ms", dt.as_secs_f64() * 1e3);
}

// ---------------------------------------------------------------------------
// Numeric helpers (also used by traversal)
// ---------------------------------------------------------------------------

/// Tiny additive term to avoid division by zero in 1/sqrt(r2).
pub(crate) const R2_TINY: f64 = f64::MIN_POSITIVE;
pub(crate) const MIN_SOFTENING: f64 = 0.0;

#[inline]
pub(crate) fn inv_r_from_r2(r2: f64) -> f64 {
    let s2 = r2 + R2_TINY;
    1.0 / s2.sqrt()
}

#[inline]
pub(crate) fn inv_r_and_inv_r3_from_r2(r2: f64) -> (f64, f64) {
    let s2 = r2 + R2_TINY;
    let inv_r = 1.0 / s2.sqrt();
    let inv_r2 = inv_r * inv_r;
    let inv_r3 = inv_r2 * inv_r;
    (inv_r, inv_r3)
}

// ---------------------------------------------------------------------------
// Tree data structures
// ---------------------------------------------------------------------------

/// BH payload for a single node (total mass + center-of-mass).
#[derive(Clone, Copy)]
pub struct NodeBh {
    pub com: [f64; 3],
    pub mass: f64,
}

/// A single octree node.
#[derive(Clone)]
pub struct Node {
    pub center: [f64; 3],
    pub half_size: f64,
    /// Cached (2*half_size)^2 for fast opening-criterion checks.
    pub size2: f64,
    /// Octant children indices into the nodes vec, or `None` for a leaf.
    pub children: Option<[usize; 8]>,
    /// Particle indices in this node's subtree.
    pub indices: Vec<usize>,
}

// ---------------------------------------------------------------------------
// Octree
// ---------------------------------------------------------------------------

/// A spatial octree over 3-D points, with optional masses and softenings.
///
/// This is the central data structure for tree-based gravity solvers.
/// It owns the particle data and all precomputed node-level payloads
/// (BH center-of-mass, per-node max softening, multipole moments).
#[derive(Clone)]
pub struct Octree {
    pub positions: Vec<[f64; 3]>,
    pub masses: Option<Vec<f64>>,
    /// Optional per-particle softening length (same length as positions).
    pub softenings: Option<Vec<f64>>,
    pub nodes: Vec<Node>,

    // ---- treewalk link caches ----
    /// For each node, first existing child index (or `usize::MAX`).
    pub first_subnode: Vec<usize>,
    /// For each node, next sibling to visit after finishing this subtree.
    pub next_branch: Vec<usize>,

    // ---- node-level payloads ----
    pub bh: Option<Vec<NodeBh>>,
    pub multipoles: Option<MultipoleMoments>,
    /// Per-node maximum softening length.
    pub hmax: Option<Vec<f64>>,

    // ---- configuration ----
    pub multipole_order: u8,
    pub leaf_capacity: usize,
    pub kernel: KernelKind,
}

impl Octree {
    // ------------------------------------------------------------------
    // Construction
    // ------------------------------------------------------------------

    fn bbox_of_points(pts: &[[f64; 3]]) -> ([f64; 3], f64) {
        let mut minp = [f64::INFINITY; 3];
        let mut maxp = [f64::NEG_INFINITY; 3];
        for p in pts {
            for i in 0..3 {
                if p[i] < minp[i] {
                    minp[i] = p[i];
                }
                if p[i] > maxp[i] {
                    maxp[i] = p[i];
                }
            }
        }
        let center = [
            (minp[0] + maxp[0]) / 2.0,
            (minp[1] + maxp[1]) / 2.0,
            (minp[2] + maxp[2]) / 2.0,
        ];
        let mut half: f64 = 0.0;
        for i in 0..3 {
            half = half.max((maxp[i] - minp[i]) / 2.0);
        }
        if half == 0.0 {
            half = 1e-6;
        }
        (center, half)
    }

    /// Build an Octree from owned data vectors.
    ///
    /// If `masses` is `None`, unit mass is assumed by all downstream
    /// computations.  `softenings` must have the same length as `positions`
    /// when provided.  `multipole_order` is clamped to at most 5.
    pub fn from_owned(
        positions: Vec<[f64; 3]>,
        masses: Option<Vec<f64>>,
        softenings: Option<Vec<f64>>,
        leaf_capacity: usize,
        multipole_order: u8,
        kernel: KernelKind,
    ) -> Self {
        let t_all = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };

        let t0 = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };
        let (center, half) = Octree::bbox_of_points(&positions);
        if let Some(t0) = t0 {
            log_timing("octree.bbox", t0.elapsed());
        }
        let n = positions.len();
        if let Some(ref hs) = softenings {
            assert_eq!(hs.len(), n, "softenings length must match positions length");
        }

        let t0 = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };
        let mut tree = Octree {
            positions,
            masses,
            softenings,
            nodes: Vec::new(),
            first_subnode: Vec::new(),
            next_branch: Vec::new(),
            bh: None,
            multipoles: None,
            hmax: None,
            multipole_order,
            leaf_capacity: leaf_capacity.max(1),
            kernel,
        };
        if let Some(t0) = t0 {
            log_timing("octree.init", t0.elapsed());
        }

        let indices: Vec<usize> = (0..n).collect();
        let root = tree.make_node(center, half, indices);
        tree.nodes.push(root);

        let t0 = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };
        tree.build_recursive(0);
        if let Some(t0) = t0 {
            log_timing("octree.build_recursive", t0.elapsed());
        }

        let t0 = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };
        tree.build_treewalk_links();
        if let Some(t0) = t0 {
            log_timing("octree.build_treewalk_links", t0.elapsed());
        }

        if let Some(t_all) = t_all {
            log_timing("octree.from_owned.total", t_all.elapsed());
        }
        tree
    }

    // ------------------------------------------------------------------
    // Tree building helpers
    // ------------------------------------------------------------------

    fn make_node(&mut self, center: [f64; 3], half_size: f64, indices: Vec<usize>) -> Node {
        let s = half_size * 2.0;
        Node {
            center,
            half_size,
            size2: s * s,
            children: None,
            indices,
        }
    }

    fn subdivide_node(&mut self, node_idx: usize) {
        let (center, half, parent_indices) = {
            let n = &mut self.nodes[node_idx];
            let center = n.center;
            let half = n.half_size;
            let parent_indices = std::mem::take(&mut n.indices);
            (center, half, parent_indices)
        };
        let mut child_indices: [usize; 8] = [usize::MAX; 8];
        let mut buckets: [Vec<usize>; 8] = Default::default();
        {
            for &pi in &parent_indices {
                let p = self.positions[pi];
                let mut oct = 0usize;
                if p[0] >= center[0] {
                    oct |= 1;
                }
                if p[1] >= center[1] {
                    oct |= 2;
                }
                if p[2] >= center[2] {
                    oct |= 4;
                }
                buckets[oct].push(pi);
            }
        }
        for oct in 0..8 {
            if buckets[oct].is_empty() {
                continue;
            }
            let mut child_center = center;
            let offset = half / 2.0;
            child_center[0] += if (oct & 1) != 0 { offset } else { -offset };
            child_center[1] += if (oct & 2) != 0 { offset } else { -offset };
            child_center[2] += if (oct & 4) != 0 { offset } else { -offset };
            let child = self.make_node(child_center, offset, std::mem::take(&mut buckets[oct]));
            let idx = self.nodes.len();
            self.nodes.push(child);
            child_indices[oct] = idx;
        }
        self.nodes[node_idx].children = Some(child_indices);
    }

    fn build_recursive(&mut self, node_idx: usize) {
        let should_subdivide = {
            let n = &self.nodes[node_idx];
            n.indices.len() > self.leaf_capacity
        };
        if !should_subdivide {
            return;
        }
        self.subdivide_node(node_idx);
        if let Some(children) = self.nodes[node_idx].children {
            for &c in &children {
                if c == usize::MAX {
                    continue;
                }
                self.build_recursive(c);
            }
        }
    }

    // ------------------------------------------------------------------
    // Treewalk link cache
    // ------------------------------------------------------------------

    pub(crate) fn build_treewalk_links(&mut self) {
        let n = self.nodes.len();
        self.first_subnode = vec![usize::MAX; n];
        self.next_branch = vec![usize::MAX; n];

        fn rec(nodes: &[Node], first: &mut [usize], next: &mut [usize], node_idx: usize) {
            let Some(children) = nodes[node_idx].children else {
                return;
            };

            let mut last: Option<usize> = None;
            for c in children {
                if c == usize::MAX {
                    continue;
                }
                if first[node_idx] == usize::MAX {
                    first[node_idx] = c;
                }
                if let Some(prev) = last {
                    next[prev] = c;
                }
                last = Some(c);
            }
            if let Some(last_child) = last {
                next[last_child] = next[node_idx];
            }
            for c in children {
                if c == usize::MAX {
                    continue;
                }
                if nodes[c].children.is_some() {
                    rec(nodes, first, next, c);
                }
            }
        }

        // Root's next branch is the end-of-traversal marker.
        self.next_branch[0] = usize::MAX;
        let nodes_ref: &[Node] = &self.nodes;
        rec(nodes_ref, &mut self.first_subnode, &mut self.next_branch, 0);
    }

    // ------------------------------------------------------------------
    // Convenience builder (backward-compatible with old Tree3D::build)
    // ------------------------------------------------------------------

    /// Convenience: build an Octree from borrowed slices with default settings
    /// (no softening, Plummer kernel).  For full control use `from_owned`.
    pub fn build(
        positions: &[[f64; 3]],
        masses: Option<&[f64]>,
        leaf_capacity: usize,
        multipole_order: u8,
    ) -> Self {
        let mut tree = Octree::from_owned(
            positions.to_vec(),
            masses.map(|m| m.to_vec()),
            None,
            leaf_capacity,
            multipole_order,
            KernelKind::Plummer,
        );
        tree.build_mass_payload();
        tree
    }

    // ------------------------------------------------------------------
    // Payload building
    // ------------------------------------------------------------------

    #[inline]
    pub fn bh(&self) -> &Vec<NodeBh> {
        self.bh
            .as_ref()
            .expect("BH payload not initialized; call build_mass_payload() before gravity queries")
    }

    pub fn set_softenings(&mut self, softenings: Option<Vec<f64>>) {
        if let Some(ref hs) = softenings {
            assert_eq!(hs.len(), self.positions.len());
        }
        self.softenings = softenings;
    }

    pub fn set_kernel(&mut self, kernel: KernelKind) {
        self.kernel = kernel;
    }

    pub fn set_masses(&mut self, masses: Option<Vec<f64>>) {
        self.masses = masses;
    }

    fn build_bh_payload(&self) -> Vec<NodeBh> {
        let masses_opt = self.masses.as_deref();
        let mut bh = vec![
            NodeBh {
                mass: 0.0,
                com: [0.0; 3]
            };
            self.nodes.len()
        ];

        for idx in (0..self.nodes.len()).rev() {
            let mut mass = 0.0f64;
            let mut com = [0.0f64; 3];
            let node = &self.nodes[idx];

            if node.children.is_none() {
                if !node.indices.is_empty() {
                    if let Some(masses) = masses_opt {
                        for &pi in &node.indices {
                            let p = self.positions[pi];
                            let m = masses[pi];
                            mass += m;
                            com[0] += p[0] * m;
                            com[1] += p[1] * m;
                            com[2] += p[2] * m;
                        }
                    } else {
                        for &pi in &node.indices {
                            let p = self.positions[pi];
                            mass += 1.0;
                            com[0] += p[0];
                            com[1] += p[1];
                            com[2] += p[2];
                        }
                    }
                    if mass > 0.0 {
                        com[0] /= mass;
                        com[1] /= mass;
                        com[2] /= mass;
                    }
                }
            } else if let Some(children) = node.children {
                for &c in &children {
                    if c == usize::MAX {
                        continue;
                    }
                    let child_bh = &bh[c];
                    if child_bh.mass == 0.0 {
                        continue;
                    }
                    mass += child_bh.mass;
                    com[0] += child_bh.com[0] * child_bh.mass;
                    com[1] += child_bh.com[1] * child_bh.mass;
                    com[2] += child_bh.com[2] * child_bh.mass;
                }
                if mass > 0.0 {
                    com[0] /= mass;
                    com[1] /= mass;
                    com[2] /= mass;
                }
            }

            bh[idx] = NodeBh { mass, com };
        }

        bh
    }

    fn build_hmax_payload(&self) -> Option<Vec<f64>> {
        let hs = self.softenings.as_deref()?;
        let mut hmax = vec![0.0f64; self.nodes.len()];

        for idx in (0..self.nodes.len()).rev() {
            let node = &self.nodes[idx];
            if node.children.is_none() {
                let mut m = 0.0f64;
                for &pi in &node.indices {
                    m = m.max(hs[pi].max(MIN_SOFTENING));
                }
                hmax[idx] = m;
            } else if let Some(children) = node.children {
                let mut m = 0.0f64;
                for &c in &children {
                    if c == usize::MAX {
                        continue;
                    }
                    m = m.max(hmax[c]);
                }
                hmax[idx] = m;
            }
        }
        Some(hmax)
    }

    pub fn build_mass_payload(&mut self) {
        let t_all = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };

        let t0 = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };
        let payload = self.build_bh_payload();
        self.bh = Some(payload);
        if let Some(t0) = t0 {
            log_timing("octree.build_bh_payload", t0.elapsed());
        }

        let t0 = if timing_enabled() {
            Some(Instant::now())
        } else {
            None
        };
        self.hmax = self.build_hmax_payload();
        if let Some(t0) = t0 {
            log_timing("octree.build_hmax_payload", t0.elapsed());
        }

        if self.multipole_order > 0 {
            let t0 = if timing_enabled() {
                Some(Instant::now())
            } else {
                None
            };
            let payload = self.build_multipole_payload();
            self.multipoles = Some(payload);
            if let Some(t0) = t0 {
                log_timing("octree.build_multipole_payload", t0.elapsed());
            }
        }

        if let Some(t_all) = t_all {
            log_timing("octree.build_mass_payload.total", t_all.elapsed());
        }
    }

    fn build_multipole_payload(&self) -> MultipoleMoments {
        let masses_opt = self.masses.as_deref();
        let bh = self
            .bh
            .as_ref()
            .expect("BH payload not initialized; call build_mass_payload() before building multipoles");
        let order = self.multipole_order.min(5);
        let mut moments = vec![MultipoleMoment::zero(); self.nodes.len()];

        for idx in (0..self.nodes.len()).rev() {
            let node = &self.nodes[idx];
            let node_bh = &bh[idx];
            if node_bh.mass == 0.0 {
                continue;
            }

            if node.children.is_none() {
                if node.indices.is_empty() {
                    continue;
                }
                let center = node_bh.com;
                let m = MultipoleMoment::from_points(
                    &self.positions,
                    masses_opt,
                    &node.indices,
                    center,
                    order,
                );
                moments[idx] = m;
            } else if let Some(children) = node.children {
                let center = node_bh.com;
                let mut acc = MultipoleMoment::zero();
                for &c in &children {
                    if c == usize::MAX {
                        continue;
                    }
                    let child_bh = &bh[c];
                    if child_bh.mass == 0.0 {
                        continue;
                    }
                    let shift = [
                        center[0] - child_bh.com[0],
                        center[1] - child_bh.com[1],
                        center[2] - child_bh.com[2],
                    ];
                    let translated = translate_multipole(&moments[c], shift, order);
                    acc.add_assign(&translated);
                }
                moments[idx] = acc;
            }
        }

        MultipoleMoments::from_full(moments, order)
    }
}

File: cpp/gravity/octree.cpp
#include "gravity/octree.hpp"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <limits>
#include <stdexcept>
#include <utility>

namespace gravity {

namespace {

void build_treewalk_links_rec(const std::vector<Node>& nodes,
                              std::vector<size_t>& first,
                              std::vector<size_t>& next,
                              size_t node_idx) {
    const Node& node = nodes[node_idx];
    if (!node.children.has_value()) {
        return;
    }
    const auto& children = *node.children;

    bool has_last = false;
    size_t last = NO_INDEX;
    for (size_t c : children) {
        if (c == NO_INDEX) {
            continue;
        }
        if (first[node_idx] == NO_INDEX) {
            first[node_idx] = c;
        }
        if (has_last) {
            next[last] = c;
        }
        last = c;
        has_last = true;
    }
    if (has_last) {
        next[last] = next[node_idx];
    }
    for (size_t c : children) {
        if (c == NO_INDEX) {
            continue;
        }
        if (nodes[c].children.has_value()) {
            build_treewalk_links_rec(nodes, first, next, c);
        }
    }
}

} // namespace

std::pair<Vec3, double> Octree::bbox_of_points(const std::vector<Vec3>& pts) {
    double minp[3] = {std::numeric_limits<double>::infinity(),
                      std::numeric_limits<double>::infinity(),
                      std::numeric_limits<double>::infinity()};
    double maxp[3] = {-std::numeric_limits<double>::infinity(),
                      -std::numeric_limits<double>::infinity(),
                      -std::numeric_limits<double>::infinity()};
    for (const Vec3& p : pts) {
        for (int i = 0; i < 3; ++i) {
            if (p[i] < minp[i]) {
                minp[i] = p[i];
            }
            if (p[i] > maxp[i]) {
                maxp[i] = p[i];
            }
        }
    }
    Vec3 center((minp[0] + maxp[0]) / 2.0,
                (minp[1] + maxp[1]) / 2.0,
                (minp[2] + maxp[2]) / 2.0);
    double half = 0.0;
    for (int i = 0; i < 3; ++i) {
        half = std::max(half, (maxp[i] - minp[i]) / 2.0);
    }
    if (half == 0.0) {
        half = 1e-6;
    }
    return {center, half};
}

Octree Octree::from_owned(std::vector<Vec3> positions,
                          std::optional<std::vector<double>> masses,
                          std::optional<std::vector<double>> softenings,
                          size_t leaf_capacity,
                          unsigned char multipole_order,
                          KernelKind kernel) {
    auto bbox = bbox_of_points(positions);
    Vec3 center = bbox.first;
    double half = bbox.second;

    size_t n = positions.size();
    if (softenings.has_value() && softenings->size() != n) {
        throw std::invalid_argument("softenings length must match positions length");
    }

    Octree tree;
    tree.positions = std::move(positions);
    tree.masses = std::move(masses);
    tree.softenings = std::move(softenings);
    tree.multipole_order = multipole_order;
    tree.leaf_capacity = (leaf_capacity < 1) ? 1 : leaf_capacity;
    tree.kernel = kernel;

    std::vector<size_t> indices(n);
    for (size_t i = 0; i < n; ++i) {
        indices[i] = i;
    }
    Node root = tree.make_node(center, half, std::move(indices));
    tree.nodes.push_back(std::move(root));

    tree.build_recursive(0);
    tree.build_treewalk_links();

    return tree;
}

Node Octree::make_node(Vec3 center, double half_size, std::vector<size_t> indices) {
    double s = half_size * 2.0;
    Node node;
    node.center = center;
    node.half_size = half_size;
    node.size2 = s * s;
    node.children = std::nullopt;
    node.indices = std::move(indices);
    return node;
}

void Octree::subdivide_node(size_t node_idx) {
    Node& n = nodes[node_idx];
    Vec3 center = n.center;
    double half = n.half_size;
    std::vector<size_t> parent_indices = std::move(n.indices);

    std::array<size_t, 8> child_indices;
    child_indices.fill(NO_INDEX);
    std::array<std::vector<size_t>, 8> buckets;

    for (size_t pi : parent_indices) {
        const Vec3& p = positions[pi];
        size_t oct = 0;
        if (p.x >= center.x) {
            oct |= 1;
        }
        if (p.y >= center.y) {
            oct |= 2;
        }
        if (p.z >= center.z) {
            oct |= 4;
        }
        buckets[oct].push_back(pi);
    }

    for (size_t oct = 0; oct < 8; ++oct) {
        if (buckets[oct].empty()) {
            continue;
        }
        Vec3 child_center = center;
        double offset = half / 2.0;
        child_center.x += (oct & 1) ? offset : -offset;
        child_center.y += (oct & 2) ? offset : -offset;
        child_center.z += (oct & 4) ? offset : -offset;
        Node child = make_node(child_center, offset, std::move(buckets[oct]));
        size_t idx = nodes.size();
        nodes.push_back(std::move(child));
        child_indices[oct] = idx;
    }

    nodes[node_idx].children = child_indices;
}

void Octree::build_recursive(size_t node_idx) {
    bool should_subdivide = nodes[node_idx].indices.size() > leaf_capacity;
    if (!should_subdivide) {
        return;
    }
    subdivide_node(node_idx);
    if (nodes[node_idx].children.has_value()) {
        // Copy by value: the recursive build_recursive(c) below push_backs into
        // `nodes`, which may reallocate it and invalidate a reference into it.
        const auto children = *nodes[node_idx].children;
        for (size_t c : children) {
            if (c == NO_INDEX) {
                continue;
            }
            build_recursive(c);
        }
    }
}

void Octree::build_treewalk_links() {
    size_t n = nodes.size();
    first_subnode.assign(n, NO_INDEX);
    next_branch.assign(n, NO_INDEX);

    // Root's next branch is the end-of-traversal marker.
    next_branch[0] = NO_INDEX;
    build_treewalk_links_rec(nodes, first_subnode, next_branch, 0);
}

Octree Octree::build(const std::vector<Vec3>& positions,
                     const double* masses,
                     size_t leaf_capacity,
                     unsigned char multipole_order) {
    std::optional<std::vector<double>> masses_opt;
    if (masses != nullptr) {
        masses_opt = std::vector<double>(positions.size());
        std::copy(masses, masses + positions.size(), masses_opt->begin());
    }
    Octree tree = Octree::from_owned(positions, std::move(masses_opt), std::nullopt,
                                     leaf_capacity, multipole_order, KernelKind::Plummer);
    tree.build_mass_payload();
    return tree;
}

const std::vector<NodeBh>& Octree::bh_payload() const {
    if (!bh.has_value()) {
        throw std::runtime_error(
            "BH payload not initialized; call build_mass_payload() before gravity queries");
    }
    return *bh;
}

void Octree::set_softenings(std::optional<std::vector<double>> softenings) {
    if (softenings.has_value() && softenings->size() != positions.size()) {
        throw std::invalid_argument("softenings length must match positions length");
    }
    this->softenings = std::move(softenings);
}

void Octree::set_kernel(KernelKind kernel) {
    this->kernel = kernel;
}

void Octree::set_masses(std::optional<std::vector<double>> masses) {
    this->masses = std::move(masses);
}

std::vector<NodeBh> Octree::build_bh_payload() const {
    const double* masses_ptr = masses.has_value() ? masses->data() : nullptr;
    std::vector<NodeBh> bh(nodes.size());
    for (size_t i = 0; i < bh.size(); ++i) {
        bh[i].com = Vec3(0.0, 0.0, 0.0);
        bh[i].mass = 0.0;
    }

    for (std::ptrdiff_t idx = static_cast<std::ptrdiff_t>(nodes.size()) - 1; idx >= 0; --idx) {
        double mass = 0.0;
        Vec3 com(0.0, 0.0, 0.0);
        const Node& node = nodes[static_cast<size_t>(idx)];

        if (!node.children.has_value()) {
            if (!node.indices.empty()) {
                if (masses_ptr != nullptr) {
                    for (size_t pi : node.indices) {
                        const Vec3& p = positions[pi];
                        double m = masses_ptr[pi];
                        mass += m;
                        com.x += p.x * m;
                        com.y += p.y * m;
                        com.z += p.z * m;
                    }
                } else {
                    for (size_t pi : node.indices) {
                        const Vec3& p = positions[pi];
                        mass += 1.0;
                        com.x += p.x;
                        com.y += p.y;
                        com.z += p.z;
                    }
                }
                if (mass > 0.0) {
                    com.x /= mass;
                    com.y /= mass;
                    com.z /= mass;
                }
            }
        } else {
            for (size_t c : *node.children) {
                if (c == NO_INDEX) {
                    continue;
                }
                const NodeBh& child_bh = bh[c];
                if (child_bh.mass == 0.0) {
                    continue;
                }
                mass += child_bh.mass;
                com.x += child_bh.com.x * child_bh.mass;
                com.y += child_bh.com.y * child_bh.mass;
                com.z += child_bh.com.z * child_bh.mass;
            }
            if (mass > 0.0) {
                com.x /= mass;
                com.y /= mass;
                com.z /= mass;
            }
        }

        bh[static_cast<size_t>(idx)].mass = mass;
        bh[static_cast<size_t>(idx)].com = com;
    }

    return bh;
}

std::optional<std::vector<double>> Octree::build_hmax_payload() const {
    if (!softenings.has_value()) {
        return std::nullopt;
    }
    const std::vector<double>& hs = *softenings;
    std::vector<double> hmax(nodes.size(), 0.0);

    for (std::ptrdiff_t idx = static_cast<std::ptrdiff_t>(nodes.size()) - 1; idx >= 0; --idx) {
        const Node& node = nodes[static_cast<size_t>(idx)];
        if (!node.children.has_value()) {
            double m = 0.0;
            for (size_t pi : node.indices) {
                m = std::max(m, std::max(hs[pi], MIN_SOFTENING));
            }
            hmax[static_cast<size_t>(idx)] = m;
        } else {
            double m = 0.0;
            for (size_t c : *node.children) {
                if (c == NO_INDEX) {
                    continue;
                }
                m = std::max(m, hmax[c]);
            }
            hmax[static_cast<size_t>(idx)] = m;
        }
    }

    return hmax;
}

void Octree::build_mass_payload() {
    std::vector<NodeBh> payload = build_bh_payload();
    bh = std::move(payload);

    hmax = build_hmax_payload();

    if (multipole_order > 0) {
        MultipoleMoments payload = build_multipole_payload();
        multipoles = std::move(payload);
    }
}

MultipoleMoments Octree::build_multipole_payload() const {
    const double* masses_ptr = masses.has_value() ? masses->data() : nullptr;
    if (!bh.has_value()) {
        throw std::runtime_error(
            "BH payload not initialized; call build_mass_payload() before building multipoles");
    }
    const std::vector<NodeBh>& bh_ref = *bh;
    unsigned char order = (multipole_order < 5) ? multipole_order : 5;
    std::vector<MultipoleMoment> moments(nodes.size(), MultipoleMoment::zero());

    for (std::ptrdiff_t idx = static_cast<std::ptrdiff_t>(nodes.size()) - 1; idx >= 0; --idx) {
        const Node& node = nodes[static_cast<size_t>(idx)];
        const NodeBh& node_bh = bh_ref[static_cast<size_t>(idx)];
        if (node_bh.mass == 0.0) {
            continue;
        }

        if (!node.children.has_value()) {
            if (node.indices.empty()) {
                continue;
            }
            Vec3 center = node_bh.com;
            MultipoleMoment m = MultipoleMoment::from_points(positions, masses_ptr,
                                                              node.indices, center, order);
            moments[static_cast<size_t>(idx)] = m;
        } else {
            Vec3 center = node_bh.com;
            MultipoleMoment acc = MultipoleMoment::zero();
            for (size_t c : *node.children) {
                if (c == NO_INDEX) {
                    continue;
                }
                const NodeBh& child_bh = bh_ref[c];
                if (child_bh.mass == 0.0) {
                    continue;
                }
                Vec3 shift = center - child_bh.com;
                MultipoleMoment translated = translate_multipole(moments[c], shift, order);
                acc.add_assign(translated);
            }
            moments[static_cast<size_t>(idx)] = acc;
        }
    }

    return multipole_moments_from_full(std::move(moments), order);
}

} // namespace gravity
